### PR TITLE
fuzz: Extend addrman fuzz test with deserialize

### PR DESCRIPTION
**Pull request by maflcko**
Requested on IRC:

```
[18:01] <vasild> => I think there is a good chance fuzzing addrman unserialize will find more bugs
[18:04] <sipa> definitely

### CODE REVIEW

**Comment by MarcoFalke** [comment]
https://github.com/bitcoin/bitcoin/files/6840256/crash-be26535b1ec81883f57c0e947e29c19efed8d3da.log:

```
$ FUZZ=addrman ./src/test/fuzz/fuzz ./crash-be26535b1ec81883f57c0e947e29c19efed8d3da.log 
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 1355821193
INFO: Loaded 1 modules   (229077 inline 8-bit counters): 229077 [0x558717fba0f8, 0x558717ff1fcd), 
INFO: Loaded 1 PC tables (229077 PCs): 229077 [0x558717ff1fd0,0x558718370d20), 
./src/test/fuzz/fuzz: Running 1 inputs 1 time(s) each.
Running: ./crash-be26535b1ec81883f57c0e947e29c19efed8d3da.log
fuzz: ./addrman.h:261: void CAddrMan::Serialize(Stream &) const [Stream = CDataStream]: Assertion `nIds != nNew' failed.

File: src/test/fuzz/addrman.cpp [hunk 1]
```diff
@@ -38,6 +38,16 @@ FUZZ_TARGET_INIT(addrman, initialize_addrman)
...
     addr_man.MakeDeterministic(ConsumeUInt256(fuzzed_data_provider));
+    if (fuzzed_data_provider.ConsumeBool()) {
+        const std::vector<uint8_t> serialized_data{ConsumeRandomLengthByteVector(fuzzed_data_provider)};
+        CDataStream ds(serialized_data, SER_DISK, INIT_PROTO_VERSION);
+        const auto ser_version{fuzzed_data_provider.ConsumeIntegral<int32_t>()};
+        ds.SetVersion(ser_version);
+        try {
+            ds >> addr_man;
+        } catch (const std::ios_base::failure&) {
+        }
```
Comment (vasild) [inline src/test/fuzz/addrman.cpp:49]: I think that `addr_man` is not in a usable state if an exception is thrown (i.e. it is in an undefined state). Below we call its methods and expect they would work. Maybe:
```suggestion
        try {
            ds >> addr_man;
        } catch (const std::ios_base::failure&) {
            addr_man.Clear();
        }
```

**Review by vasild — COMMENTED** [review]
Concept ACK

The unserialize itself is fuzzed in:

https://github.com/bitcoin/bitcoin/blob/d3474b8df2f973e9b9142c0b64505a8a78bcb292/src/test/fuzz/deserialize.cpp#L201-L204

What this test adds is calling addrman methods after it is unserialized from fuzzed data. This kind of simulates a disk corruption. I think it may result in the same addr:port being in "new" and "tried" or other "interesting" situations that do not occur from normal operation.

File: src/test/fuzz/addrman.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/test/fuzz/addrman.cpp]: Good catch. Fixed.

**Review by vasild — APPROVED** [review]
ACK fa740252d2dcc272204b745906785fe26e9a0b77

**Review by jonatack — COMMENTED** [review]
Tested ACK fa740252d2dcc272204b745906785fe26e9a0b77

File: src/test/fuzz/addrman.cpp [hunk 2]
```diff
@@ -38,6 +38,17 @@ FUZZ_TARGET_INIT(addrman, initialize_addrman)
...
+        const auto ser_version{fuzzed_data_provider.ConsumeIntegral<int32_t>()};
+        ds.SetVersion(ser_version);
+        try {
+            ds >> addr_man;
+        } catch (const std::ios_base::failure&) {
+            addr_man.Clear();
+        }
+    }
     if (fuzzed_data_provider.ConsumeBool()) {
         addr_man.m_asmap = ConsumeRandomLengthBitVector(fuzzed_data_provider);
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/test/fuzz/addrman.cpp:53]: True, this works for me.
```diff
@@ -38,6 +38,12 @@ FUZZ_TARGET_INIT(addrman, initialize_addrman)
     SetMockTime(ConsumeTime(fuzzed_data_provider));
     CAddrManDeterministic addr_man;
     addr_man.MakeDeterministic(ConsumeUInt256(fuzzed_data_provider));
+    if (fuzzed_data_provider.ConsumeBool()) {
+        addr_man.m_asmap = ConsumeRandomLengthBitVector(fuzzed_data_provider);
+        if (!SanityCheckASMap(addr_man.m_asmap)) {
+            addr_man.m_asmap.clear();
+        }
+    }
     if (fuzzed_data_provider.ConsumeBool()) {
         const std::vector<uint8_t> serialized_data{ConsumeRandomLengthByteVector(fuzzed_data_provider)};
         CDataStream ds(serialized_data, SER_DISK, INIT_PROTO_VERSION);
@@ -49,12 +55,6 @@ FUZZ_TARGET_INIT(addrman, initialize_addrman)
             addr_man.Clear();
         }
     }
-    if (fuzzed_data_provider.ConsumeBool()) {
-        addr_man.m_asmap = ConsumeRandomLengthBitVector(fuzzed_data_provider);
-        if (!SanityCheckASMap(addr_man.m_asmap)) {
-            addr_man.m_asmap.clear();
-        }
-    }
```

[the remark this answers is not in the document]
Comment (vasild) [reply src/test/fuzz/addrman.cpp:53]: yes

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/test/fuzz/addrman.cpp]: Done

**Comment by jonatack** [comment]
ACK aaaa9c6019790a1a21a7b4ef01693ac9390ae6d0 per `git diff fa74025 aaaa9c6`

**Comment by MarcoFalke** [comment]
@vasild I am seeing there is a conflict. Do you have a preference which one should be merged first?

**Review by vasild — APPROVED** [review]
ACK aaaa9c6019790a1a21a7b4ef01693ac9390ae6d0

> @vasild I am seeing there is a conflict. Do you have a preference which one should be merged first?

No.

**Comment by MarcoFalke** [comment]
Ok, merging the one with the nicer commit-hash first :sweat_smile:

**Pull request merged**